### DISCUSSION TITLE
Opening large files results in severe hangs or the app refusing to render any content in the window.

### DISCUSSION THREAD

**Discussion by denil-ct** [krzyzanowskim/STTextView#12] [Questions]
When I tried to open large files in code edit, the app became unresponsive and stuck with a transparent window. On profiling the app the issue seemed to lie in STTextView. So, tried the same file with the demo app in this repo and got the same issue. I have attached a sample file that reproduces the issue.

[lipsum.txt](https://github.com/krzyzanowskim/STTextView/files/10538972/lipsum.txt)

![image](https://user-images.githubusercontent.com/95201442/215568928-4d8d8cf6-fe5c-4ee9-8ca1-b76d9dd90aba.png)

**Comment by krzyzanowskim** [comment]
confirmed. It's my fault (intentional temporary workaround "until resolve problem". I've got a fix for that but not sure what's broken now ¯\_(ツ)_/¯


https://user-images.githubusercontent.com/758033/215585373-eab8f1fa-0f88-4789-a930-a13fe38d0eec.mp4

**Reply by denil-ct** [reply]
Awesome, great to hear. Let me know if you need any help in testing this.

**Comment by krzyzanowskim** [comment]
The "fix" is on `main` branch. It's not full fix tho. the Line number ruler brings the performance down, and also is broken now. Don't use it if you us `STLineNumberRulerView` and stick to latest release unstead

**Comment by krzyzanowskim** [comment]
Improved performance is on `main` branch now. It seems to work just fine at first glance. @denil-ct please take a look before I make it a release.

https://user-images.githubusercontent.com/758033/216732467-eab833df-5862-4c7f-8495-3ed36cdd9570.mov

**Comment by denil-ct** [comment]
@krzyzanowskim it seems the initial issue has been fixed and the scroll performance has improved greatly.
But it seems to be triggering some constraints exception just before reaching the end of the document.
`The window has been marked as needing another Update Constraints in Window pass, but it has already had more Update Constraints in Window passes than there are views in the window.`

**Reply by krzyzanowskim** [reply]
that's caused by annotation subviews. not happening when remove annofations while scrolling. I don't have a solution to that yet.

**Reply by denil-ct** [reply]
Might be because of parallel updation of subviews?? 
Wrapping the `updateLineAnnotations` block within  a `Task` block seems to fix the problem.

**Reply by krzyzanowskim** [reply]
unlikely. I just pushed a fix to that that looks like do the job

**Reply by denil-ct** [reply]
I tried filtering on `viewportRange` earlier, but faced some issues, like after scrolling for a bit, the reported text range was wrong. The start location will be greater than the end location. Once that happens if you go back to your annotations they won't be visible, since the annotation locations will be outside the range.
Did you face any issues like that?

**Reply by krzyzanowskim** [reply]
not with the current version



https://user-images.githubusercontent.com/758033/216784299-30bce3c2-a7ad-4383-b58b-6321168e3f44.mov

**Reply by denil-ct** [reply]
Try two finger scroll to the end of the document and then back to the first line. Scrubbing the scrollbar does not seem to produce this issue.

**Reply by krzyzanowskim** [reply]
rrright. applied workaround for that. Hopefully that solves it

**Reply by denil-ct** [reply]
Yep, seems to be fixed 🎉
